ssz/src/main/java/tech/pegasys/artemis/util/SSZTypes/Bitlist.java

```diff
@@ -13,41 +13,63 @@
 
 package tech.pegasys.artemis.util.SSZTypes;
 
-import static java.util.Objects.isNull;
+import static com.google.common.base.Preconditions.checkElementIndex;
 
-import java.util.Arrays;
+import java.util.ArrayList;
+import java.util.BitSet;
+import java.util.List;
+import java.util.Objects;
 import java.util.stream.IntStream;
 import org.apache.tuweni.bytes.Bytes;
 
 public class Bitlist {
 
-  private byte[] byteArray;
-  private long maxSize;
+  private final BitSet data;
+  private final int size;
+  private final long maxSize;
 
   public Bitlist(int arraySize, long maxSize) {
-    this.byteArray = new byte[arraySize];
+    this.size = arraySize;
+    this.data = new BitSet(arraySize);
     this.maxSize = maxSize;
   }
 
   public Bitlist(Bitlist bitlist) {
-    this.byteArray = new byte[bitlist.getByteArray().length];
-    for (int i = 0; i < bitlist.getByteArray().length; i++) {
-      this.byteArray[i] = bitlist.getByteArray()[i];
-    }
+    this.size = bitlist.size;
+    this.data = (BitSet) bitlist.data.clone();
     this.maxSize = bitlist.getMaxSize();
   }
 
-  public Bitlist(byte[] bitlist, long maxSize) {
-    this.byteArray = bitlist;
+  private Bitlist(int size, BitSet data, long maxSize) {
+    this.size = size;
+    this.data = data;
     this.maxSize = maxSize;
   }
 
   public void setBit(int i) {
-    this.byteArray[i] = 1;
+    checkElementIndex(i, size);
+    data.set(i);
+  }
+
+  public boolean getBit(int i) {
+    checkElementIndex(i, size);
+    return data.get(i);
+  }
+
+  public boolean intersects(Bitlist other) {
+    return data.intersects(other.data);
   }
 
-  public int getBit(int i) {
-    return byteArray[i];
+  public List<Integer> getAllSetBits() {
+    final List<Integer> setBits = new ArrayList<>();
+    for (int i = data.nextSetBit(0); i > 0; i = data.nextSetBit(i + 1)) {
```

Comment:
Are you accounting for the 0 index being set here? 
```suggestion
    for (int i = data.nextSetBit(0); i >= 0; i = data.nextSetBit(i + 1)) {
```